plugins/in_logcat/in_logcat.c

```diff
@@ -0,0 +1,409 @@
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2015-2022 The Fluent Bit Authors
+ *  Copyright (C) 2005-2017 The Android Open Source Project
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "in_logcat.h"
+#include "logprint.h"
+#include "entities.h"
+
+#include <fluent-bit/flb_input.h>
+#include <fluent-bit/flb_input_plugin.h>
+#include <fluent-bit/flb_config.h>
+#include <fluent-bit/flb_pack.h>
+
+#include <msgpack.h>
+
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <stdarg.h>
+
+static int output_log_entry(const AndroidLogEntry * logEntry,
+                            struct flb_logcat *ctx);
+static int connect_to_logd_read_socket(struct flb_input_instance *in,
+                                       struct flb_config *config,
+                                       struct flb_logcat *ctx);
+static int input_collector_delete(int coll_id, struct flb_input_instance *in);
+static struct flb_input_collector *get_collector(int id,
+                                                 struct flb_input_instance
+                                                 *in);
+
+int output_error_as_log(struct flb_logcat *ctx, const char *fmt, ...)
+{
+    char buf[256];
+    va_list args;
+    va_start(args, fmt);
+    vsnprintf(buf, sizeof(buf), fmt, args);
+    va_end(args);
+
+    AndroidLogEntry entry;
+    entry.message = buf;
+    entry.tag = "in_logcat";
+    entry.tagLen = 10;
+    entry.messageLen = strlen(buf);
+    entry.pid = getpid();
+    entry.uid = 0;
+    entry.priority = ANDROID_LOG_ERROR;
+    entry.tid = getpid();
+
+    time_t t = time(NULL);
+    entry.tv_sec = t;
+    entry.tv_nsec = t * NS_PER_SEC;
+
+    output_log_entry(&entry, ctx);
+    return 0;
+}
+
+static int output_log_entry(const AndroidLogEntry * logEntry,
+                            struct flb_logcat *ctx)
+{
+    char buf[4096];
+    size_t line_length;
+    bool ext_buf_used = false;
+    char *line = formatLogLine(buf, sizeof(buf), logEntry, &line_length);
+    if (!line) {
```

Comment:
`NULL` should be explicitly checked for, for some integer based types the rule is they can be used at face value (ie. `if (variable_of_type_we_know_only_zero_means_false) { ... }`)but `NULL` has to be explicitly compared (ie. `if (x == NULL) { ... }`)

I personally think all types should be explicitly compared and I advocate for having constants at the left side of the comparison but that last part is not part of the style and thus not allowed =(